Add support for the Gregorian Calendar availableFormats

gregtatum opened the pull request:
This PR adds support for the `availableFormats` key to the `DateTimeFormats`, in order to provide support for skeleton matching. It is part of the dependencies for Issue #481. I'm splitting out this piece from the rest of the work, as it is fairly self contained, and it's what the rest of the work will be based upon. It has a fairly large code diff, as it's generating lots of test data, so it should be helpful to get it in place first.

This is ready for review.

(If anyone is curious, my draft PR for all of the components bag work is available in #479, which includes things like Skeleton.)

gregtatum commented on resources/testdata/data/json/dates/gregory@1/ar-EG.json:
```diff
@@ -123,10 +123,250 @@
+        "medium": "{1}, {0}",
+        "short": "{1}, {0}"
+      },
+      "available_formats": [
```
I'd prefer this to be formatted like:

```
"available_formats": [
  ["Bh", "h B"],
  ["Bhm", "h:mm B"],
  ["Bhms", "h:mm:],
  ...
]
```

I don't see any field attributes for making something more condensed in the serialization, so I think it could only be accomplished with a custom serializer. This doesn't really seem worth it.

zbraniecki replied:
I think this may be your solution - https://crates.io/crates/serde-tuple-vec-map - we already use it in ICU4X, so it wouldn't add to dependency chain.

gregtatum replied:
Oof, of course I used that to read in the CLDR data, but then forgot I could use it to re-export the data. Thanks for the tip.

gregtatum commented on resources/testdata/data/json/dates/gregory@1/ar-EG.json:
```diff
@@ -123,10 +123,250 @@
+        "medium": "{1}, {0}",
+        "short": "{1}, {0}"
+      },
+      "available_formats": [
```
I was trying to decide on the best representation for this data, as it's a key pair value. Making it a simple tuple seemed like the best choice here to me.

Ultimately, the key will be converted from `Cow<'static, str>` to a `struct Skeleton` that can then be used for skeleton matching. I'm not 100% sure if it's useful to provide these as `Cow`s, or what the ultimate design of the memory management for providers is. However, this is following the existing examples on how this data is being read in, so I'm assuming it's compatible.

My understanding of the data flow is
* The serialized JSON is read in.
* This gets deserialized into the `DateTimeFormatsV1` structure.
* `Skeleton`s are generated from the `DateTimeFormatsV1`
* Skeleton matching happens with the `Skeleton` struct.

zbraniecki replied:
Yeah, I think that for now we can keep the key as `Cow<str>`, and separately once we have more complete data structures we can experiment with shifting from stringified representations to serializing/deserializing structs and see what it does to perf/mem/size.

gregtatum commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +193,17 @@ pub mod gregory {
+            feature = "provider_serde",
+            derive(serde::Serialize, serde::Deserialize)
+        )]
+        pub struct DateTimeFormatsV1 {
```
At what point do we need to care about versioning conflicts, and will we have a testing strategy for catching them? I'm assuming for now we're considering this format to be unstable.

sffc replied:
Don't worry about stability of data structs until we reach v1.

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +193,17 @@ pub mod gregory {
+            pub style_patterns: StylePatternsV1,
+
+            #[cfg_attr(feature = "provider_serde", serde(with = "tuple_vec_map"))]
+            pub available_formats: Vec<(Cow<'static, str>, Cow<'static, str>)>,
```
If you put cows in a data struct, you should add a lifetime parameter to the data struct instead of using `'static`.  Alternatively, you can use TinyStr (for skeletons) or SmallStr (for patterns) and obviate the need for a lifetime parameter.

It looks like the skeletons are always, or almost always, 8 ASCII characters or shorter, in which case TinyStr8 would work very well.  If you want to allow skeletons to be longer than that, consider TinyStrAuto.

gregtatum replied:
> If you put cows in a data struct, you should add a lifetime parameter to the data struct instead of using 'static. 

This is where a lot of my confusion over the memory model for data providers really stems. All of the other strings are using the `'static` lifetime parameter, so I'm not sure if there is any difference between that usage, and this usage. I was honestly a bit surprised this code compiled.

```
    symbols!(months, [Cow<'static, str>; 12]);

    symbols!(weekdays, [Cow<'static, str>; 7]);

    symbols!(
        day_periods {
            am: Cow<'static, str>,
            pm: Cow<'static, str>,
            noon: Option<Cow<'static, str>>,
            midnight: Option<Cow<'static, str>>,
        }
    );
```

I investigated the source of the Cows, which appears to be #256. It seems like this was initially landed to get the data in tree, with the idea that we would optimize in the future. Similarly I'm a bit nervous about spending a lot of time optimizing at the beginning, but let's start with a decent choice. I think there are different trade-offs in this data provider format, but I'd be happy to try the the non-heap/inline solutions first, and then measure the memory afterwards.

Should I file an issue to add lifetime parameters to the `Cow` structures that are already here? If we're not adding new ones, it seems like the existing ones should be addressed at some point.

gregtatum replied:
Here is a breakdown of pattern lengths across all of the CLDR.

<img width="602" alt="image" src="https://user-images.githubusercontent.com/1588648/107390751-9ec4e880-6abd-11eb-931d-9bd8ecb9cf9f.png">

gregtatum replied:
Re-posting this with correct byte size, instead of character count:

<img width="602" alt="image" src="https://user-images.githubusercontent.com/1588648/107393532-873b2f00-6ac0-11eb-9e88-365127232138.png">

gregtatum replied:
Per the style guide:

> - If lifetime parameters are allowed, use `&'a str`.
> - If lifetime parameters are not allowed, use one of:
>   - [TinyStr](https://github.com/zbraniecki/tinystr) if the string is ASCII-only.
>   - [SmallString](https://crates.io/crates/smallstr) for shorter strings, with a stack size ∈ {8, 12, 16, 20} that fits at least 99% of cases.
>   - `Cow<'static, str>`  for longer strings.

The largest suggested stack size of 20 for this data set is: (aside: I'm not sure if stack size in the style guide is including the data required by by the SmallString struct itself beyond just the data)

20 bytes: 96.5%
29 bytes: 98.8%
30 bytes: 99.5%

Which, according to the style guide suggests we use `Cow<'static, str>`.

sffc replied:
I need to update the style guide. As of my recent changes to the data provider trait, cows can and should have real lifetime parameters. I thought I updated all existing structs when I did that PR but I guess I missed this one hiding in a macro. You don't have to worry about migrating in this PR; I'll follow up in #257.

About data types: I'm more talking about skeletons than patterns. I believe that patterns are going to be long enough that you need a cow. However, I think the keys (skeletons) might be a good just case for TinyStr.

sffc commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -130,6 +130,21 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
+                medium: other.medium.get_pattern().clone(),
+                short: other.short.get_pattern().clone(),
+            },
+            available_formats: other.available_formats.0.clone(),
```
You don't have to copy the available_formats literally from CLDR into ICU4X.  Is there any pre-processing you can do to optimize them?

sffc commented on resources/testdata/data/json/dates/gregory@1/ar-EG.json:
```diff
@@ -123,10 +123,73 @@
+        "MMMEd": "E، d MMM",
+        "MMMMd": "d MMMM",
+        "MMMMEd": "E، d MMMM",
+        "MMMMW-count-zero": "الأسبوع W من MMMM",
+        "MMMMW-count-one": "الأسبوع W من MMMM",
+        "MMMMW-count-two": "الأسبوع W من MMMM",
+        "MMMMW-count-few": "الأسبوع W من MMMM",
+        "MMMMW-count-many": "الأسبوع W من MMMM",
+        "MMMMW-count-other": "الأسبوع W من MMMM",
```
Do you have the plural rule selection implemented?

gregtatum replied:
No I don't, and I haven't really thought through this problem. Do you have recommendations here? Currently I discard them in the `Skeleton` struct construction. I don't really have a mental model of how these will be used.

gregtatum replied:
Ok, I'm going to try to change the representation here to handle these. I see what's going on with them now.

sffc replied:
If you don't want to feature-creep your work, open an issue to figure these out later, and in the mean time, copy only the "other" form into the data.

sffc reviewed (changes requested):
I would like to see the code using these resources before we add the resources to the data model.

gregtatum commented:
Thanks for the review!

> I would like to see the code using these resources before we add the resources to the data model.

That seems fair enough. I would like to work incrementally, so I may investigate ways to separate out different sections still.

sffc commented:
> Thanks for the review!
> 
> > I would like to see the code using these resources before we add the resources to the data model.
> 
> That seems fair enough. I would like to work incrementally, so I may investigate ways to separate out different sections still.

It's good to keep this as a separate PR; I'd just like to review the code PR first so I can get an idea of how this data is being used.  I imagine that there may be additional changes you may want to make once we see how the data is used.

gregtatum commented:
I updated the PR to address the review feedback so far. I still need to rebase my bigger work on top of these changes, so I may tweak the design a bit more, however this incorporates the review feedback so far.

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -194,16 +195,90 @@ pub mod gregory {
+            pub pattern: Cow<'static, str>,
+            #[cfg_attr(
+                feature = "provider_serde",
+                serde(skip_serializing_if = "Option::is_none")
```
You need to also make this dependent on the serialize_none feature, which is necessary for bincode.  Find some boilerplate from elsewhere and copy it in here.

We should probably add a CI check that fails if bincode doesn't round-trip.

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +194,91 @@ pub mod gregory {
+        // currently implement Serialize or Deserialize traits.
+        //
+        // pub type SkeletonsV1 = Vec<(TinyStr16, SmallVec<[SkeletonPatternV1; 1]>)>,
+        pub type SkeletonsV1 = Vec<(Cow<'static, str>, SmallVec<[SkeletonPatternV1; 1]>)>;
```
Suggestion: instead of using `SmallVec<[SkeletonPatternV1; 1]>`, consider an `enum` with two variants: a single pattern, and a TupleVecMap of multiple patterns keyed by plural form.

gregtatum commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -140,7 +143,80 @@ impl From<&cldr_json::DateTimeFormats> for gregory::patterns::DateTimeFormatsV1
+                        skeletons
+                            .last_mut()
+                            .and_then(|(last_skeleton, last_patterns)| {
+                                debug_assert!(
+                                    *last_skeleton <= unique_skeleton,
+                                    "The skeletons must be sorted."
+                                );
```
This is an assertion for the sort that I perform earlier, so if it fails it means there is a programming error, not a data error. I believe keeping it as a `debug_assert` is the proper choice here, as it will only fail if someone removes that sort. The data from CLDR is not correctly sorted alphabetically. I could make this more clear with an additional comment.

https://github.com/unicode-org/icu4x/pull/480/files/63983715c4af1caed5877d7f4a61ba66da7a4677..a5e20def6e06d671114dfe51ca831ecfbd148467#diff-4313b054e5b8d61c9fba3243afb5070dc55e8caae89d8922d8970cf1bd97ef3bR149-R153

sffc replied:
Thanks; didn't see the sort earlier.

gregtatum commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -132,6 +132,9 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
 
 impl From<&cldr_json::DateTimeFormats> for gregory::patterns::DateTimeFormatsV1 {
     fn from(other: &cldr_json::DateTimeFormats) -> Self {
+        use gregory::patterns::{CountV1, SkeletonPatternV1, SkeletonsV1};
```
I have a few thoughts here:

A: I was wondering about dependencies when I wrote it. I looked into using `PluralCategory` originally, but didn't understand the strategy for inter-dependencies between the various component crates. I could use the `PluralCategory` directly, but this would add `icu_plurals` as a dependency to `icu_datetime`. Is this OK?

B: I'm a bit concerned with sharing serializable representations with code outside of the provider. The `V1` signals a social contract to maintain the serialized represented. If I naively pulled in `PluralCategory`, then this gives no signal that it's used as a serializable source, and care should be taken when changing its representation. Perhaps this is fine here since this is a fairly standardized representation. However, is there still a risk of this assumption changing? This type of thing was a source of bugs when I maintained the Firefox Profiler.

C: I read through [data-pipeline.md](https://github.com/unicode-org/icu4x/blob/master/docs/design/data_pipeline.md) again, but I'm still not 100% sure on the guarantees here. Here in the provider, every struct has an associated version.

* How granular is this version number?
* Does this mean that individual pieces of the serialization can change? For example, what happens if I make a breaking change to `CountV1`, does that mean every struct that contains a `CountV1` needs to be bumped versions?
* Do we not care yet to define this breakdown, and label everything as V1 until something else is needed?
* Can we have something with a V2 key, but contain V1 structs?
* What happens when we have a breaking change?

zbraniecki replied:
> A: I was wondering about dependencies when I wrote it. I looked into using PluralCategory originally, but didn't understand the strategy for inter-dependencies between the various component crates. I could use the PluralCategory directly, but this would add icu_plurals as a dependency to icu_datetime. Is this OK?

We could have it be an optional dependency and have skeletons be behind a feature (on by default, but this would allow you to build datetime with just styles for smaller footprint!)

sffc replied:
I mean, icu_datetime is going to have to depend on icu_plurals anyway in order to operate on this data, right?

In any case, don't have any problem with adding this dependency because icu_datetime is higher-level than icu_plurals.  It will also need to eventually gain a dependency on icu_numbers.

sffc replied:
On your other questions:

B: This would assume `PluralCategory` is serializable as an enum.  If it's not, then I would implement serde on that enum over in icu_plurals behind that crate's `serde` feature.  Then the stability guarantee comes from PluralCategory's serde form.

C: Let's not worry about data struct versioning until v1.  These are going to be interesting questions to discuss and answer when we actually have to start worrying about our stability guarantees.  Until that time, let's not worry about it much.

zbraniecki replied:
> I mean, icu_datetime is going to have to depend on icu_plurals anyway in order to operate on this data, right?

What I'm saying is that `icu_datetime` that is only working with dateStyle/timeStyle is a usable date/time formatting API that may *not* pull in `icu_plurals`, am I correct?
`icu_plurals` become important when `skeletons` come into play.

sffc replied:
Just re-reading the style guide on crate features:

https://github.com/unicode-org/icu4x/blob/master/docs/process/style_guide.md#when-to-add-crate-features--suggested

> When adding enhancements to an ICU4X component, introduce features as a way for the end user to control the code size of their compilation as follows:
>
> 1. If the enhancement adds a new crate dependency, it should be behind a feature.
> 2. If the enhancement contains code that is not considered best practice, for example if it is mainly used for debugging diagnostics or development, then it should be behind a feature.

So it looks like this may fall under case 1.  My hope is that dead code elimination would take care of this one, but I have no problem with putting the skeleton handling stuff behind a feature flag.

gregtatum replied:
Given the conversation here, I'm doubting the use of storing the string data in memory at all. It would be fairly easy to implement `Serialize` and `Deserialize` for `Field` and `PatternItem`. Then the data could use something like the following shape:

```rs
pub type SkeletonTupleV1 = (
    SmallVec<[fields::Field; 5]>,
    SmallVec<[Vec<pattern::PatternItem>; 1]>,
);
```

Then all of the pattern matching machinery can take a lifetime parameter, and do all of its work via reference, without worrying about owning or copying any of this data. The data could be serialized back into the standard UTS 35 representations for strings, which is convenient and well-documented, but the in-memory representation would be exactly how we want to use it for the actual data processing.

The processing step can handle pre-sorting the information to make sure it's in the most efficient and direct form for skeleton matching. The skeleton matching machinery could then be a wrapper with a lifetime over this data representation.

sffc replied:
I am very much in favor of pre-processing CLDR strings such that the data stored in ICU4X DataProvider needs minimal parsing and processing at runtime.  I think it's okay if it incurs a small data size penalty as well.

gregtatum commented:
It turns out the skeleton data in the CLDR has identical patterns for all of the different plural variants (except "other"). I'm thinking it will be better and simpler to land this code only with the assumption that there is a 1:1 relationship between skeleton to pattern.

The only skeletons that have multiple patterns have at _most_ one variant. This includes for plural rules, where the count-one, count-two, etc variants are identical. The only plural rule that has anything different is the `count-other` field.

Here is the list of patterns that have variants for a given a skeleton.

| locale | skeleton | pattern |
| --- | --- | --- |
| en-CA | Md | MM-dd |
| en-CA | Md-alt-variant | d/M |
| en-CA | MEd | E, MM-dd |
| en-CA | MEd-alt-variant | E, d/M |
| en-CA | MMdd | MM-dd |
| en-CA | MMdd-alt-variant | dd/MM |
| en-CA | yM | y-MM |
| en-CA | yM-alt-variant | M/y |
| en-CA | yMd | y-MM-dd |
| en-CA | yMd-alt-variant | d/M/y |
| en-CA | yMEd | E, y-MM-dd |
| en-CA | yMEd-alt-variant | E, d/M/y |
| fil | MMMMW-count-one | 'ika'-W 'linggo' 'ng' MMMM |
| fil | MMMMW-count-other | 'linggo' W 'ng' MMMM |
| fil | yw-count-one | 'ika'-w 'linggo' 'ng' Y |
| fil | yw-count-other | 'linggo' w 'ng' Y |
| hy | MMMMW-count-one | MMMM W-ին շաբաթ |
| hy | MMMMW-count-other | MMMM W-րդ շաբաթ |
| hy | yw-count-one | Y թ․ w-ին շաբաթ |
| hy | yw-count-other | Y թ․ w-րդ շաբաթ |
| pcm | yw-count-one | 'Wik' w 'fọ' Y |
| pcm | yw-count-other | 'Wiik' w 'fọ' Y |
| ps | MMMMW-count-one | اونۍ‘ W د MMMM‘ |
| ps | MMMMW-count-other | اونۍ W د MMMM |
| ps-PK | MMMMW-count-one | اونۍ‘ W د MMMM‘ |
| ps-PK | MMMMW-count-other | اونۍ W د MMMM |

This list is small enough, that I don't think it's worth handling. In addition, we are not accepting free-form pattern input to match against, but rather specific options bags for the components. I don't think there is an easy way to prefer one variant over the other. This will simplify the initial implementation, and then we can re-visit the decision if we actually do need a way to access these variants.

[I determined this with this script.](https://gist.github.com/gregtatum/775379b393da798c7be172f31f954f3d)

I also did a [spot](https://github.com/unicode-org/cldr/blob/bcc9fbd0546078d5d8e80be29df1c444268fec6b/common/main/ar.xml#L1984-L1989) check of the CLDR XML data to ensure that the JSON data was accurate, and it appears so. (I checked this locally more thoroughly.)

sffc commented:
> This list is small enough, that I don't think it's worth handling.

The list being small shouldn't by itself justify low-level design decisions like this one if they reduce i18n quality. For example, it looks like the plural forms produce actually different results for "pcm".

> In addition, we are not accepting free-form pattern input to match against, but rather specific options bags for the components. I don't think there is an easy way to prefer one variant over the other.

I'm also not sure what to do about variants that aren't tied to a Unicode extension keyword.

I know you want to check this in, but I'd like to discuss this question at the meeting, because I think it's an important decision that could serve as precedent for future scenarios. So can we talk about it in Thursday?

gregtatum commented:
> The list being small shouldn't by itself justify low-level design decisions like this one if they reduce i18n quality.

First off, I agree that just because a list is small, that it shouldn't be reason to reduce the quality of the overall output. I'm focusing on delivering a correct solution, but also keeping an eye towards limiting scope of this initial implementation.

My understanding was that the `count-other` was equivalent to `alt-variant`, but digging further, I realize that this is not the case. Looking into the `plurals.json` test data I see the following definition:

```
      "pcm": {
        "pluralRule-count-one": "i = 0 or n = 1 @integer 0, 1 @decimal 0.0~1.0, 0.00~0.04",
        "pluralRule-count-other": " @integer 2~17, 100, 1000, 10000, 100000, 1000000, … @decimal 1.1~2.6, 10.0, 100.0, 1000.0, 10000.0, 100000.0, 1000000.0, …"
      },
```

This leads me to realize that effectively for "pcm" the `count-other` is actually the _plural_ variant. So I agree, this should be handled here, otherwise it would be a low quality result.

> I'm also not sure what to do about variants that aren't tied to a Unicode extension keyword.

I think for this first check-in, we should ignore them, and then file a follow-up to discuss them and come to a decision on them.

> I know you want to check this in, but I'd like to discuss this question at the meeting, because I think it's an important decision that could serve as precedent for future scenarios. So can we talk about it in Thursday?

I'd be happy to discuss further if you feel the need, but I think we may be in agreement now that I realize my misunderstanding of the `count-other` case. In summary:

* I will handle the `-count-other` case
* I will not handle the `-alt-variant` case, and file a follow-up issue to discuss.

sffc commented:
That sounds good.  Alternatively, if you want to reduce the scope, I see that the plural forms are only used in the "week of year" and "month of year" patterns; you could just omit those patterns, since there is no components bag that triggers them, and make a note to follow up in #488.

gregtatum commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +207,380 @@ pub mod gregory {
+        )]
+        pub struct DateTimeFormatsV1 {
+            pub style_patterns: StylePatternsV1,
+            pub skeletons: Vec<SkeletonTupleV1>,
```
tuple_vec_map does not appear to work when you've wrapped the tuple in another struct.

gregtatum commented:
I've updated this PR with the existing feedback. There are 4 new commits listed here. Locally, I rewrote my history while I was working, but used a merge strategy to update this PR. You can look at the total code diff, or look at it on a per-commit basis. This would be Part 1, and Part 2 of the new commits:

* Merge in master
* Part 1: Directly serialize patterns and skeletons in the CLDR and prov…
* Part 2: Re-compute the testdata for skeletons
* Merge branch 'components-bag' (early part) into available-formats

sffc commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -107,8 +110,174 @@ impl From<Vec<PatternItem>> for Pattern {
+}
+
+#[cfg(feature = "provider_serde")]
+impl<'de> Deserialize<'de> for Pattern {
```
Okay, so here's the thing.

The CLDR date pattern syntax is almost definitely *not* the best way to store patterns in the data provider.  I've been saying this for a while, including in my latest update to data_pipeline.md.  It is designed for human readability, not for performance.  What you're doing here is to make the CLDR date pattern syntax the de-facto standard for the ICU4X data model.  Adding custom serde comes with costs; we should only do it if we're very confident about the data model.

Why can't you just store the parsed array of pattern tokens in the data provider?

sffc replied:
Main issue: #523 

We decided to keep the exotic type in the data struct.

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +207,380 @@ pub mod gregory {
+                    }
+                }
+
+                serializer.serialize_str(&string)
```
For skeletons and patterns: you need to toggle to a machine-readable format based on is_human_readable.  See how @Manishearth did it in TinyStr: https://github.com/zbraniecki/tinystr/pull/34

gregtatum commented:
If you want to look at what's just changed, it's part 3 and part 4. I merged in the latest `main`, and then merged in my bigger commit stack.

4265117 - Part 1: Directly seralize patterns and skeletons in the CLDR and prov…
2884ccd - Part 2: Re-compute the testdata for skeletons
0989370 - Part 3: Add bincode serialization support for Patterns
240ed1b - Part 4: Add impl Display for Pattern

gregtatum commented:
I've got some CI errors. Please hold off on reviewing until I get a chance to look at them, which I'll probably do tomorrow.

gregtatum commented:
I resorted to a force push, as things were getting really messy in here, and CI was not passing. This is now ready for re-review.

sffc commented on components/datetime/tests/fixtures/tests/patterns.json:
```diff
@@ -0,0 +1,24 @@
+[
```
FYI: `patterns.bin` is empty.  Delete?

gregtatum replied:
It's a binary file, so it's showing up as empty in the github UI, but it 2.5kb.

<img width="608" alt="image" src="https://user-images.githubusercontent.com/1588648/111483685-0563a480-8703-11eb-9bde-d4c8bb912d42.png">

```
~/dev/icu4x on icu4x/components-bag at 09:28:07
➤ xxd ./components/datetime/tests/fixtures/tests/patterns.bin
00000000: 1600 0000 0000 0000 1800 0000 0000 0000  ................
00000010: 0100 0000 0000 0000 0000 0000 0200 0000  ................
00000020: 0000 0000 0000 0000 7000 0000 0000 0000  ........p.......
00000030: 0700 0000 0000 0000 0000 0000 0300 0000  ................
00000040: 0000 0000 0000 0000 0100 0000 0200 0000  ................
00000050: 0000 0000 2c20 0000 0000 0100 0000 0000  ...., ..........
00000060: 0000 0000 0000 0100 0000 0100 0000 0000  ................
```

It's used in `components/datetime/tests/pattern_serialization.rs`

sffc commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -115,6 +115,83 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
+                    //   "MMMMW-count-two"
+                    //   unique_skeleton: "MMMMW"
+                    //   variant_text: "-count-two"
+                    let chars = skeleton_str.chars();
```
It's nicer to use `skeleton_str.split('-')` and the pull items off that iterator.  You could also consider using itertools' `next_tuple()` and expand it into a tuple like I do in #541.

gregtatum replied:
I opted for the split implementation, but not the `next_tuple` as it would add dependencies and include an assumption that variant text would always have a fixed number of splits.

gregtatum replied:
This is: dd688d8efee45053a50789db64d90eb753fe22a2

sffc commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -115,6 +115,83 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
+                            Err(err) => match err {
+                                // Ignore unimplemented fields for now.
+                                SkeletonFieldsV1Error::SymbolUnimplemented(_) => continue,
+                                SkeletonFieldsV1Error::SymbolUnknown(symbol) => panic!(
```
I have not yet formed a strong opinion on whether we should `panic!` in transformer code.  I've usually been returning errors from functions instead of panicking, but I realize that maybe it's acceptable to panic in transformer code, because it helps better identify where the error originated.  But I do think we should be consistent.  @Manishearth?

sffc replied:
Follow-up: #574

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +207,380 @@ pub mod gregory {
+        )]
+        pub struct DateTimeFormatsV1 {
+            pub style_patterns: StylePatternsV1,
+            pub skeletons: Vec<SkeletonTupleV1>,
```
I think this is a textbook example of something that should be `LiteMap<Skeleton, Pattern>`

gregtatum replied:
I believe LiteMap was created after I wrote this, but I went ahead and updated it to use LiteMap.

gregtatum replied:
This is: e12a4bb6761a4f7355791f292bc0711f89df3eea

sffc commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -115,6 +115,83 @@ impl From<&cldr_json::StylePatterns> for gregory::patterns::StylePatternsV1 {
+                }
+
+                // Sort the skeletons deterministically by their canonical sort order.
+                skeleton_tuples.sort_by(|a, b| a.0.compare_canonical_order(&b.0));
```
If you use LiteMap here, the sorting will be done automatically.

gregtatum replied:
This required some trait implementations, but LiteMap is now handling the sorting.

gregtatum replied:
This is: e12a4bb6761a4f7355791f292bc0711f89df3eea

sffc commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -374,6 +451,26 @@ pub(self) mod cldr_json {
+
+    #[derive(PartialEq, Clone, Debug, Deserialize)]
+    pub struct AvailableFormats(
+        #[serde(with = "tuple_vec_map")] pub(crate) Vec<(Cow<'static, str>, Cow<'static, str>)>,
```
Since the CLDR JSON is only ever created from `serde::from_reader`, these strings will never be static.  Therefore, I think we should start using plain `String` in CLDR JSON structs (only in CLDR JSON, not in ICU4X JSON!)

gregtatum replied:
I agree, however I would like this to be a follow-up, since this code is a change to the existing practice. I filed #558

sffc reviewed (changes requested):
Phew!  Looking better.  Here is my next round of feedback.

gregtatum commented on components/datetime/src/fields/length.rs:
```diff
@@ -3,12 +3,16 @@
 
 #[derive(Debug, PartialEq, Clone, Copy)]
+#[cfg_attr(
+    feature = "provider_serde",
```
> I think the "serde" feature should enable serde for library types, and "provider_serde" should enable it for data structs.

Ah, that kind of separation makes sense.

I filed #556 to document the features, so that we can have some shared understanding of the intent of various features. I've found myself struggling with how to think about them, especially in the meta crate context.

> On the other hand, this enum does not get exported, so maybe it's okay to put it on "provider_serde". Thoughts?

I like being able to write code with how it's being used, so I'm inclined to keep it provider_serde. Although this choice should probably be documented.

gregtatum commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -107,8 +122,161 @@ impl From<Vec<PatternItem>> for Pattern {
     }
 }
 
+impl fmt::Display for Pattern {
```
I'm a bit confused by this feedback, as the `Writeable` trait needs to know the size of the output ahead of time. In the example of `WelcomeMessage`, this is easy to determine as it's `8 + self.name.len()`. However, here this is deriving a string. In order to know the length, the string must be generated first.

It looks like the closest example in the code may be the `FixedDecimal` `Writeable`. This is duplicating the branching logic but determining the string size rather than the string itself. Another similar example is the `FormattedDateTime`. However, there is a TODO as I'm guessing this will be quite complex as well.

To me this does not seem like a good use case for `Writeable`, as the determination of string size is not a particularly trivial operation, and this is not hot code.

I was somewhat hesitant to convert this over to `fmt::Display`, as the need for this code was to convert it to a string, while the `fmt::Display` is not particularly in the current requirements.

The action I'm going to take here is to add a note about Writeable for future readers with my analysis.

sffc replied:
My analysis suggests that computing the length of the string can fairly substantially enhance performance, and it's not usually as hard to do as you think.  But, this particular example is low priority and not a good use of time.  So just leave the comment and it is fine.

gregtatum commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -107,8 +122,161 @@ impl From<Vec<PatternItem>> for Pattern {
+                        string.push(ch);
+                    }
+                }
+                PatternItem::Literal(literal) => {
```
Is this code checked in? I would prefer not to prematurely abstract code.

zbraniecki replied:
It's going to be abstracted with my patterns utility which handles CLDR quotes.

zbraniecki replied:
I don't think we should block on it here.

gregtatum commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -107,8 +122,161 @@ impl From<Vec<PatternItem>> for Pattern {
 
+impl fmt::Display for Pattern {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        let mut string = String::new();
```
This was because I didn't know how to use the formatter correctly, as the examples I looked up in the Rust docs don't give examples of using multiple formatter calls, and this is probably my first time implementing a `fmt::Display` trait.

One suggestion I have for this type of feedback is that it's difficult to understand the intent of this kind of comment. I would suggest taking the "why do you" out of the feedback, and instead forming it more of a neutral direct statement or suggestion, e.g. "Pushing onto an intermediate string is less efficient than `formatter.write_str()` here, is this intentional?".

sffc replied:
I didn't know if there was a reason to append to a string instead of directly to the formatter, and assumed you have have done so intentionally. I don't consider myself to have a greater sense of Rust style than you do, which is why I asked it as a question instead of making a concrete suggestion. I don't want to specifically ask you to change something that I'm not certain about. :)

gregtatum replied:
Ah, thanks for the clarification. I tend to try to be as explicit as possible on review feedback, as I often find reviews to be sources of contention (especially working on multi-cultural teams.) I didn't interpret the comment as a real question, but more a challenge to the approach. For cases like this I usually rely on [conventional comments](https://conventionalcomments.org/). My old team used them and they were really helpful to know what kind of feedback we were receiving from each other.

Anyway, aside from the meta-review conversation. I rewrote this to write directly into the formatter, and read up a bit on things. I think it's a better solution than creating the string.

gregtatum commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -107,8 +122,161 @@ impl From<Vec<PatternItem>> for Pattern {
+        let pattern = match Pattern::from_bytes(pattern_string) {
+            Ok(p) => p,
+            Err(err) => {
+                return Err(E::custom(format!(
+                    "The pattern \"{}\" could not be parsed: {:?}",
+                    pattern_string, err
+                )));
```
Ah thanks, this created a bit of a code diff, but I think it produces better errors. I encapsulated the change in a diff.

gregtatum commented:
I'm still working through the changes, but I will push up my changes from today. Thanks for the review.

gregtatum commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +207,380 @@ pub mod gregory {
         }
+
+        #[derive(Debug, PartialEq, Clone, Default)]
+        pub struct SkeletonFieldsV1(pub SmallVec<[fields::Field; 5]>);
```
> Why are you using an exotic type for the Pattern but not for the Skeleton?

Technically a `fields::Field` is an exotic type.

In my skeleton matching I'm using the field like so:

```
pub struct Skeleton<'a> {
    fields: &'a SkeletonFieldsV1,
}
```

And then the available formats turns into:

```
pub struct AvailableFormatPattern<'a> {
    /// The skeleton is used to match against.
    skeleton: Skeleton<'a>,
    pub pattern: &'a Pattern,
}
```

See: https://github.com/unicode-org/icu4x/pull/479/files#diff-b5d51c874f2246611510523fcd93bb1b1028a21027be78ea1bb0f5c03da401d3R42

sffc replied:
My main objection here is that I want to keep the `crate::provider` module tidy.

Let me be more clear.  I would prefer if you would change what you currently have,

```rust
// skeleton.rs
pub struct Skeleton<'a> {
    fields: &'a SkeletonFieldsV1,
}

// provider/mod.rs
pub struct SkeletonFieldsV1(pub SmallVec<[fields::Field; 5]>);
pub struct SkeletonTupleV1(pub SkeletonFieldsV1, pub PatternV1);
pub struct PatternV1(pub Pattern);
pub struct DateTimeFormatsV1 {
    pub style_patterns: StylePatternsV1,
    pub skeletons: Vec<SkeletonTupleV1>,
}
```

to

```rust
// skeleton.rs
pub struct Skeleton {
    pub fields: SmallVec<[fields::Field; 5]>,
}

// provider/mod.rs
pub struct DateTimeFormatsV1 {
    pub style_patterns: StylePatternsV1,
    pub skeletons: LiteMap<Skeleton, Pattern>,
}
```

I don't *really* understand why you need to wrap `Pattern` in `PatternV1`, but you can keep that wrapper (and make a similar `SkeletonV1` wrapper) if you need it.

gregtatum replied:
Ah yes, I agree with this feedback. I made the changes.

> I don't really understand why you need to wrap Pattern in PatternV1, but you can keep that wrapper (and make a similar SkeletonV1 wrapper) if you need it.

I may want to follow-up and reconsider this.

gregtatum replied:
The primary move is here: e4ecc1acaf2be9ff2ca6ae4a156cc11c555b41a5

However, there were a few follow-up commits to deal with it.

gregtatum commented:
I found a new issue. I didn't enforce sorting of the skeleton fields from the CLDR. This was fixed in:

1584020710681b22348138453040a0618a0c4fdd

sffc commented on components/datetime/src/skeleton.rs:
```diff
@@ -0,0 +1,408 @@
+    }
+}
+
+fn compare_fields(field_a: &Field, field_b: &Field) -> Ordering {
```
Suggestion: Put this code as `impl Ord for field::Field`.  Once you do this, then I *believe* the Skeleton Ord impl should "just work" via `#[derive(Ord)]`.

gregtatum replied:
Oh nice, this simplifies this code quite a bit. I'm starting to like leaning into the trait Impls. It makes things semantic and re-usable. So far, I tend to focus on implementing unique functions to do the job at hand.

sffc commented on components/datetime/src/skeleton.rs:
```diff
@@ -0,0 +1,408 @@
+                iter.next();
+            }
+
+            if let Some(prev_field) = fields.last() {
+                in_order = in_order
+                    && prev_field.symbol.get_canonical_order() < field_symbol.get_canonical_order();
+            }
```
Suggestion: If the field is out of order, `insert` it into the SmallVec.  An easy way to do this would be to walk backward from the end of the SmallVec until you find the position where you need to insert the field.  If the skeleton is already sorted, then you always need to check just one element, and you insert it at the end.

You could also use binary search for the correct position, which is what LiteMap does under the hood.

You would then be able to get rid of `SkeletonError::FieldsOutOfOrder` and the code to handle it elsewhere.

gregtatum replied:
Ah ok, I was trying to write code that would make it impossible to accidentally use unsorted fields. I guess it doesn't really matter that much if I accept unsorted fields here as bincode is the faster path. This change fits nicely with the ordering changes.

sffc commented on components/datetime/src/provider/mod.rs:
```diff
@@ -193,5 +200,82 @@ pub mod gregory {
+                match Skeleton::try_from(skeleton_string) {
+                    Ok(skeleton) => Ok(SkeletonV1(skeleton)),
+                    Err(SkeletonError::FieldsOutOfOrder(fields)) => {
+                        // The fields were out of order, converting fields to a skeleton will ensure
+                        // that these fields are sorted. This sorting is only done in the provider.
+                        Ok(SkeletonV1(Skeleton::from(fields)))
```
Observation: Although the sorting is only done in the provider, this impl requires the sorting *code* to be shipped regardless.  If you apply my suggestion above, then you won't need this code any more.

sffc reviewed (changes requested):
Nice work!  A few nitpicks and suggestions.

gregtatum commented on components/datetime/src/skeleton.rs:
```diff
@@ -129,19 +68,46 @@ impl<'de> de::Visitor<'de> for SkeletonFieldsDeserializeVisitor {
+        while let Some(item) = seq.next_element()? {
+            items.push(item)
+        }
+        Ok(Skeleton(items))
```
Done.

Question: Do we know what kind of validation we need to do for binary sources and what kind of guarantees serde bincode provides?

gregtatum replied:
Actually, if we don't trust the source, then I'd also like to check for duplicate fields.

gregtatum replied:
Ok, this is done and with tests.

sffc replied:
Good question.  IMO, our responsibility for validation is only on the level of semantic correctness.  Bincode's own deserializer code should be able to handle correctness of types and such.  We need to perform semantic correctness validation whenever we have our own Deserialize impl, or when deserializing into a struct that has certain invariants to maintain.

gregtatum replied:
I think I agree here. I messed around with the binary representation, and serde seemed to do the correct thing when providing invalid binary data.

gregtatum commented on components/datetime/src/skeleton.rs:
```diff
@@ -189,38 +165,29 @@ impl TryFrom<&str> for Skeleton {
+        fields.sort();
         Skeleton(fields)
     }
 }
```
Done.

gregtatum commented on components/datetime/src/fields/symbols.rs:
```diff
@@ -136,6 +136,18 @@ impl From<FieldSymbol> for char {
+
+impl Ord for FieldSymbol {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.get_canonical_order().cmp(&other.get_canonical_order())
```
Done.

sffc commented on components/datetime/src/skeleton.rs:
```diff
@@ -91,6 +91,20 @@ impl<'de> de::Visitor<'de> for DeserializeSkeletonBincode {
         let mut items: SmallVec<[fields::Field; 5]> = SmallVec::new();
         while let Some(item) = seq.next_element()? {
+            if let Some(prev_item) = items.last() {
+                if prev_item > &item {
+                    return Err(de::Error::invalid_value(
+                        de::Unexpected::Other(&format!("field item out of order: {:?}", item)),
+                        &"ordered field symbols representing a skeleton",
+                    ));
+                }
+                if prev_item == &item {
+                    return Err(de::Error::invalid_value(
+                        de::Unexpected::Other(&format!("duplicate field: {:?}", item)),
+                        &"ordered field symbols representing a skeleton",
+                    ));
+                }
```
Nitpick (optional): combine error handling via `prev_item >= &item`.  Less code to read and compile for an error that shouldn't happen in normal situations.

gregtatum replied:
Since everything is green, I may fast follow with this one.

sffc reviewed (approved):
Switching my review status to Approval with one more optional nitpick
